doc/contrib/new_api/core/public.h

```diff
@@ -0,0 +1,122 @@
+#ifndef ELEKTRA_CORE_PUBLIC_H
+#define ELEKTRA_CORE_PUBLIC_H
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+// indicates special rules for constness, see comments at use-site for details
+#define ELEKTRA_CONST const
+
+typedef uint8_t ElektraNamespace;
+enum
+{
+	ELEKTRA_NS_NONE = 0,
+	ELEKTRA_NS_CASCADING = 1,
+	ELEKTRA_NS_META = 2,
+	ELEKTRA_NS_SPEC = 3,
+	ELEKTRA_NS_PROC = 4,
+	ELEKTRA_NS_DIR = 5,
+	ELEKTRA_NS_USER = 6,
+	ELEKTRA_NS_SYSTEM = 7,
+	ELEKTRA_NS_DEFAULT = 8,
+};
+
+typedef struct
+{
+	// constness follows containing struct
+	// if the containing struct is a `const ElektraName *` then this field must not be modified
+	// if the containing struct is a `ElektraName *` or a non-pointer `ElektraName` it is okay to modify
+	ELEKTRA_CONST char * name;
+	size_t size;
+	ElektraNamespace ns;
+} ElektraName;
+
+typedef struct
+{
+	const void * value;
+	size_t size;
+} ElektraValue;
+
+typedef struct ElektraEntry ElektraEntry;
+typedef struct ElektraSet ElektraSet;
+
+// 0 always indicates success, other values depend on function
+typedef int ElektraReturnCode;
+
+typedef enum
+{
+	ELEKTRA_KEY_CP_NAME = 1 << 0,
+	ELEKTRA_KEY_CP_VALUE = 1 << 1,
+	ELEKTRA_KEY_CP_META = 1 << 2,
+	ELEKTRA_KEY_CP_ALL = ELEKTRA_KEY_CP_NAME | ELEKTRA_KEY_CP_VALUE | ELEKTRA_KEY_CP_META,
+} ElektraEntryCopyFlag;
+
+void ElektraNamePushPart (ElektraName * name, const char * part);
```

Comment:
I don't think we should use stack terminology for building up a path (rather append/add/concat/join or similar). Do we even need a "pop"-equivalent?

Reply:
I used stack terminology, because we only support appending exactly one part to the end or removing exactly one part from the end. That is exactly LIFO behaviour. 

Like I said already, "append" is a good alternative for "push", but I don't see a good alternative for "pop" (*). In that case I much prefer using push/pop, a common pair, rather than e.g. append/pop, which may confuse people because it seems like a mismatch.

(*) "remove"/"delete"/etc. all convey removing something, but not that it is explicitly the last element that gets removed.

> Do we even need a "pop"-equivalent?

Well, if you can only add to a name, there would be no way to replace the current `keySetBaseName` (without rebuilding the entire name).